[api-major] Remove the `SINGLE_FILE` build target and the `PDFJS.disableWorker` option

Snuffleupagus opened the pull request:
*Please refer to the commit messages for additional details.*

This PR is an initial attempt at implementing https://mozilla.logbot.info/pdfjs/20180102#c14068811-c14068833; feedback welcome!

 - [x] Update the examples. Given the performance implications of loading the worker file in the main-thread, do we actually want to mention this in the examples or should we simply not advertise it at all?

Snuffleupagus committed:
[api-major] Remove the `SINGLE_FILE` build target

Please note that this build target, and the resulting `build/pdf.combined.js` file, is equivalent to setting the `PDFJS.disableWorker` option to `true` which is a performance footgun.

Snuffleupagus committed:
[api-major] Remove the `PDFJS.disableWorker` option

Despite this patch removing the `disableWorker` option itself, please note that we'll still fallback to loading the worker file(s) on the main-thread when running in environments without proper Web Worker support.

Furthermore it's still possible, even with this patch, to force the use of fake workers by manually loading the necessary file using a `<script>` tag on the main-thread.[1]
That way, the functionality of the now removed `SINGLE_FILE` build target and the resulting `build/pdf.combined.js` file can still be achieved simply by adding e.g. `<script src="build/pdf.worker.js"></script>` to the HTML (obviously with the path adjusted as needed).

Finally note that the `disableWorker` option is a performance footgun, and unfortunately many existing third-party examples actually use it without providing any sort of warning/justification.

---

[1] This approach is used in the default viewer, since certain kind of debugging may be easier if the code is running directly on the main-thread.

timvandermeij commented on web/app.js:
```diff
@@ -1512,6 +1513,36 @@ if (typeof PDFJSDev === 'undefined' || PDFJSDev.test('GENERIC')) {
+      };
+      (document.getElementsByTagName('head')[0] || document.body).
+        appendChild(script);
+      }
```
Nit: this bracket is misindented.

Snuffleupagus replied:
Whoops, too much copy-pasting of existing code here (and above as well); good eye!

timvandermeij commented on web/app.js:
```diff
@@ -1512,6 +1513,36 @@ if (typeof PDFJSDev === 'undefined' || PDFJSDev.test('GENERIC')) {
+      script.onload = function() {
+        resolve();
+      };
+      script.onerror = function () {
```
Nit: remove space after `function` to be consistent with the one above.

timvandermeij reviewed (commented):
This patch looks really good! Aside from the two minor comments below, I think that we should indeed not advertise disabling workers in the examples.

Snuffleupagus committed:
Remove the (unnecessary) `WorkerMessageHandler` variable from the `setupFakeWorkerGlobal()` function in the `src/display/api.js` file

Snuffleupagus commented:
/botio-linux preview

Rob--W commented on web/app.js:
```diff
@@ -1512,6 +1513,36 @@ if (typeof PDFJSDev === 'undefined' || PDFJSDev.test('GENERIC')) {
+      script.onerror = function() {
+        reject(new Error(`Cannot load fake worker at: ${script.src}`));
+      };
+      (document.getElementsByTagName('head')[0] || document.body).
```
This can be simplified to `document.head || document.body`. Only use `document.head` is only undefined in IE8-, and supported by all other browsers: https://developer.mozilla.org/en-US/docs/Web/API/Document/head

I usually do `document.head || document.documentElement` (since `<head>` should exist; if it does not, then `document.body` may also be empty, but the root element should always exist, so that's a safe fallback).

Snuffleupagus replied:
I was just copying the existing pattern used in https://github.com/mozilla/pdf.js/blob/master/web/app.js#L1515-L1534; should we simply change both them?

Rob--W replied:
You can do that in a separate commit (to keep the refactoring separate from the actual feature).

Snuffleupagus replied:
> I usually do `document.head || document.documentElement`

This suggesting has now been implemented (only in the newly added code); thank you!

Snuffleupagus replied:
> I usually do `document.head || document.documentElement`

This suggestion has now been implemented (only in the newly added code); thank you!

Rob--W commented on web/chromecom.js:
```diff
@@ -65,7 +64,9 @@ let ChromeCom = {
         'filesystem:' + location.origin + '/external/');
 
-    if (/^filesystem:/.test(file) && !PDFJS.disableWorker) {
+    const disableWorker = (window.pdfjsDistBuildPdfWorker &&
```
Under which circumstances is `window.pdfjsDistBuildPdfWorker` unset? When the worker script is loaded in the main thread, right? Is it possible for the code to reach this place before the worker script has had a chance to initiate?

In either case, please add a brief comment here to explain what this line is doing..

Snuffleupagus replied:
> Under which circumstances is window.pdfjsDistBuildPdfWorker unset? When the worker script is loaded in the main thread, right?

It's unset by default, i.e. when workers *are* being used. Hence it should only ever be defined when the `pdf.worker.js` file is loaded on the main-thread.

> Is it possible for the code to reach this place before the worker script has had a chance to initiate?

That shouldn't happen, since the `resolvePDFFile` function is being called from
https://github.com/mozilla/pdf.js/blob/fe5102a27fac7235e14334664cfc7e6dcaa04a92/web/chromecom.js#L353-L355
which in turn is called from https://github.com/mozilla/pdf.js/blob/fe5102a27fac7235e14334664cfc7e6dcaa04a92/web/app.js#L591-L596
which is called from
https://github.com/mozilla/pdf.js/blob/fe5102a27fac7235e14334664cfc7e6dcaa04a92/web/app.js#L1652-L1654
`webViewerOpenFileViaURL` is then called from
https://github.com/mozilla/pdf.js/blob/fe5102a27fac7235e14334664cfc7e6dcaa04a92/web/app.js#L1536
which is only *after* the viewer initialization code has run, see
https://github.com/mozilla/pdf.js/blob/fe5102a27fac7235e14334664cfc7e6dcaa04a92/web/app.js#L502-L504

Finally, note that we always wait on the hash parameters (among other things) to be initialized in `PDFViewerApplication.initialize`
https://github.com/mozilla/pdf.js/blob/fe5102a27fac7235e14334664cfc7e6dcaa04a92/web/app.js#L166-L172

---

> In either case, please add a brief comment here to explain what this line is doing..

Sure, will do.

*Ninja-edit:* Actually, do we still need this code-path?
This code seem to have been originally added in PR #4598, but the referenced upstream bug http://crbug.com/362061 seem to have been fixed over 3 years ago.
I'd thus assume that the affected versions of Chrome is no longer supported since quite some time, so can we (in a separate PR) remove this code-path and thus not have to worry about it here?

Rob--W replied:
I try to not break backcompat unless necessary. I should take a look at my telemetry (https://github.com/Rob--W/pdfjs-telemetry) to see whether it is viable to rip out code that supports old Chrome versions (there is lots of other code besides this line).

Snuffleupagus replied:
A comment has now been added here.

Rob--W reviewed (commented):
Just a question, how am I supposed to test PDF.js without workers in the Chrome extension build target?

Snuffleupagus commented:
> how am I supposed to test PDF.js without workers in the Chrome extension build target?

I don't know exactly how you're currently testing this configuration, so the following might be too simple, but wouldn't it be possible to do something like the following:
Set the `pdfBugEnabled` preference, and then append `disableWorker=true` to the viewer (or PDF) URL.

Rob--W commented:
> Set the `pdfBugEnabled` preference, and then append `#disableWorker=true` to the viewer (or PDF) URL.

I used to use the `#disableWorker=true` parameter, but when that was not supported by default I just put a breakpoint at the end of the library and then set `PDFJS.disableWorker=true` . Your suggestion looks a bit neater, thanks for that.

Snuffleupagus commented:
/botio test

Snuffleupagus commented:
@yurydelendik Since this PR attempts to implement the idea you proposed at https://mozilla.logbot.info/pdfjs/20180102#c14068811-c14068833, do you have time to review (or provide feedback) here?

wojtekmaj commented:
I'm an author of React-PDF and I'm concerned about this change.

How does it impact users using Webpack, who don't want to use worker? Previously,  I just set up disableWorker=true option and pdf.js requested pdf.worker.js via require() and created a fake worker. How that procedure will change? Will I need to require the worker file myself to disable worker?

Rob--W commented:
@wojtekmaj Your current use case of `PDFJS.disableWorker` has a bad impact on performance/UX. You traded performance (something that is important to users) for ease of code distribution (which is only convenient for developers).

I guess that it's technically possible to serialize the worker script and then use `URL.createObjectURL` to load a worker from a single file, but that may be broken by overly aggressive code optimizers (e.g. by having moved a function outside of the scope of the serialized function). It may also break when a strict Content Security Policy is enforced.

You should keep the PDF.js worker in a separate file, and assign the location of it to `PDFJS.workerSrc` (if it is at a non-standard location):

https://github.com/mozilla/pdf.js/blob/b6c57d90887c1a936e2638e83cf88576aea8ca36/src/display/global.js#L140-L148

The logic of worker script resolution is here:
https://github.com/mozilla/pdf.js/blob/b6c57d90887c1a936e2638e83cf88576aea8ca36/src/display/api.js#L41-L45
https://github.com/mozilla/pdf.js/blob/b6c57d90887c1a936e2638e83cf88576aea8ca36/src/display/api.js#L1205-L1218

wojtekmaj commented:
Thanks for information @Rob--W. I actually realize the benefits of service worker here and I discourage my users from using a version without worker, but I do provide them with such ability as for many beginners setting it up is an huge problem. I did struggle a lot with this myself.

Also, sometimes bundle size matters. Currently how PDF.js plays with Webpack is a mess. Sadly, unless provided with additional configuration (which I failed to figure out), Webpack includes pdf.worker.js code **twice** because of using two separate loaders on them - it's required via worker-loader in webpack.js entry file, and then *again* without using worker-loader in pdf.js itself. Because of using separate loaders Webpack is not smart enough to dedupe them.

Rob--W commented:
The code changes LGTM.

The code path for fake workers seems to not be covered by unit tests. Wouldn't it make sense to add a single test that customized `PDFJS.workerSrc` and test whether the fake worker can be loaded? Or will we rely on manual tests only for this case?

Snuffleupagus commented:
> The code path for fake workers seems to not be covered by unit tests.

Given that workers are always disabled in Node.js, we should have *implicit* fake worker test-coverage thanks to the (subset of) API unit-tests running on Travis.
https://github.com/mozilla/pdf.js/blob/fd242ad2c2191f5de32f6b158e15db8e012f9467/src/display/api.js#L54-L55
Slightly off topic for this PR, but we should see if some of the currently `pending` API tests could be enabled on Node.js/Travis since we now have [node_stream.js](https://github.com/mozilla/pdf.js/blob/master/src/display/node_stream.js) available.

> Wouldn't it make sense to add a single test that customized `PDFJS.workerSrc` and test whether the fake worker can be loaded?

While we might be able to add a Node.js-only explicit test for that, I'm not sure that we'd be able to easily test this when the unit-tests run normally (i.e. in browser) without risk of negatively impacting subsequent tests.
The reason being that once we've setup the fake workers, we'll no longer attempt to use real workers any more; refer to
https://github.com/mozilla/pdf.js/blob/fd242ad2c2191f5de32f6b158e15db8e012f9467/src/display/api.js#L1436-L1439
and
https://github.com/mozilla/pdf.js/blob/fd242ad2c2191f5de32f6b158e15db8e012f9467/src/display/api.js#L1323-L1325

> Or will we rely on manual tests only for this case?

Given the above, I'm slightly on the fence about adding a fake worker test here!

Rob--W commented:
> > The code path for fake workers seems to not be covered by unit tests.
> 
> Given that workers are always disabled in Node.js, we should have implicit fake worker test-coverage thanks to the (subset of) API unit-tests running on Travis:

I am referring to the injected script tag in your new `loadFakeWorker` function from this PR. There is probably no test coverage for that.

> Given the above, I'm slightly on the fence about trying to add a fake worker test here! Opinions?

How about a custom parameter to disable workers in the unit test? It doesn't even need to be a PDF.js API, it could also be something that does `window.Worker = null`.

Snuffleupagus commented:
> How about a custom parameter to disable workers in the unit test? It doesn't even need to be a PDF.js API, it could also be something that does window.Worker = null.

Sorry, but I really don't understand how this would help. As mentioned above, as soon as [`PDFWorker._setupFakeWorker`](https://github.com/mozilla/pdf.js/blob/fd242ad2c2191f5de32f6b158e15db8e012f9467/src/display/api.js#L1435) has executed *once* then workers become permanently disabled.

Rob--W commented:
> > How about a custom parameter to disable workers in the unit test? It doesn't even need to be a PDF.js API, it could also be something that does window.Worker = null.
>
> Sorry, but I really don't understand how this would help. As mentioned above, as soon as `PDFWorker._setupFakeWorker` has executed once then workers become permanently disabled.

When I wrote that, I had in mind that it would not be run by default until someone manually edits the URL and adds something like `&disableworker=true`. But I guess that since it is already a manual test, one can as well add a breakpoint with a debugger and manually do `window.Worker = null` for testing. Or enable debugging in the viewer and add `&disableworker=true` and see if the viewer works. So let's ignore my previous request.

One last thing: The `Worker` code path uses `getWorkerSrc()` whereas your newly introduced `<script>` tag uses `PDFJS.workerSrc` only. This could potentially result in an unexpected behavioral difference, where the Worker loads successfully, but the non-worker fails. I'll defer to your judgement on whether this should be fixed.

Snuffleupagus committed:
[api-minor] Add a (static) `PDFWorker.getWorkerSrc` method that returns the current `workerSrc`

This method returns the currently used `workerSrc`, which thus allows obtaining the fallback `workerSrc` value (e.g. when the option wasn't set by the user).

Snuffleupagus commented:
> The Worker code path uses getWorkerSrc() whereas your newly introduced <script> tag uses PDFJS.workerSrc only. This could potentially result in an unexpected behavioral difference, where the Worker loads successfully, but the non-worker fails.

That's a very good catch, thanks for spotting this!
I agree that this inconsistency seem bad, so I've addressed it by adding a new method to `PDFWorker` that allows (indirect) access to the internal `getWorkerSrc` helper function.

Snuffleupagus commented:
/botio-linux preview

Rob--W commented on src/display/api.js:
```diff
@@ -1491,6 +1502,14 @@ var PDFWorker = (function PDFWorkerClosure() {
   };
 
+  PDFWorker.getWorkerSrc = function() {
+    try {
```
Why try-catch? `getWorkerSrc` is not expected to throw (unless there is no way to determine the worker source.

Snuffleupagus replied:
While you're correct that it *shouldn't* throw, than may still happen in some cases [given this code](https://github.com/mozilla/pdf.js/blob/55e3f97aa97ad648ef955203524bbdb5593671d1/src/display/api.js#L1217)

In e.g. Node.js, where workers are disabled by default and `workerSrc` isn't set in the examples, if that function is called a (perhaps unexpected) error would be thrown. With the `getWorkerSrc` function exposed to the outside, I simply wanted to avoid "surprising" anyone using this function.

Snuffleupagus replied:
While you're correct that it *shouldn't* throw, than may still happen in some cases [given this code](https://github.com/mozilla/pdf.js/blob/55e3f97aa97ad648ef955203524bbdb5593671d1/src/display/api.js#L1217).

In e.g. Node.js, where workers are disabled by default and `workerSrc` isn't set in the examples, if that function is called a (perhaps unexpected) error would be thrown.
With the `getWorkerSrc` function now exposed to the outside, it seemed nice to avoid "surprises" for an API consumer who uses that function.

Snuffleupagus replied:
While you're correct that it *shouldn't* throw, that could still happen [given this code](https://github.com/mozilla/pdf.js/blob/55e3f97aa97ad648ef955203524bbdb5593671d1/src/display/api.js#L1217).
Hence it seemed nice to try and avoid surprises for an API consumer that uses this function not expecting an `Error` to be thrown. Are you saying that you think this is a completely unwarranted concern, and that should return the internal `getWorkerSrc` value as-is?

Rob--W replied:
Yes, I suggest to not catch the error. `workerSrc` doesn't make sense in Node.js, and when it does, we have to provide a suitable implementation instead of silently returning nothing.

Snuffleupagus replied:
I've make the requested changes; thanks for the review comments!

Rob--W commented on test/unit/api_spec.js:
```diff
@@ -405,6 +403,12 @@ describe('api', function() {
+      let workerSrc = PDFWorker.getWorkerSrc();
+      expect(typeof workerSrc).toEqual('string');
+      expect(workerSrc).toEqual(PDFJS.workerSrc);
+      done();
```
Since the test is synchronous, remove `done();` and `done` in the `function(done){`. `done` is only needed for async tests.

Snuffleupagus replied:
Done :-)

Snuffleupagus commented:
@yurydelendik Ping; since this PR implements your idea from https://mozilla.logbot.info/pdfjs/20180102#c14068811-c14068833, did you want to look at it before we land this?

Kerumen commented:
@Snuffleupagus This PR was released on `pdfjs-dist` as a patch update ([v2.0.305](https://github.com/mozilla/pdfjs-dist/tree/v2.0.305/build) to [v2.0.310](https://github.com/mozilla/pdfjs-dist/tree/v2.0.310/build)).
Unfortunately, it broke the lib react-pdf (https://github.com/wojtekmaj/react-pdf/issues/149) and maybe others. It was tagged as `api-major`, why did you release this as a patch update?

Snuffleupagus commented:
*Regarding https://github.com/mozilla/pdf.js/pull/9385#issuecomment-363030784:*

We're currently working towards the release of PDF.js version `2.0`; that work is being tracked in https://github.com/mozilla/pdf.js/projects/5. As can be seen, there's still a few outstanding things to address, which is why it's not officially released yet; the official releases are always available at https://github.com/mozilla/pdf.js/releases.

Since we wanted to be able to refactor/cleanup old code, the *major* version number was thus increased to signal that there's changes that are by design *not* backwards compatibility.
Hence upgrading from PDF.js version `1.x.y` to version `2.x.y` cannot be expected to be a drop-in replacement, without suitable adjustments to the code using the PDF.js library.

In closing, the problem seem to be twofold:
First of all, users of `pdfjs-dist` didn't pin the version number to `1.x.y` until PDF.js version `2.0` has been officially released. Secondly, some users (might have) overlooked the significance of the increased *major* version number.

Finally, the actual breaking change in `pdfjs-dist` was thus a lot earlier than stated above, more specifically in https://github.com/mozilla/pdfjs-dist/tree/v2.0.87.

Kerumen commented:
@Snuffleupagus I don't understand. The breaking change (`pdf.combined.js` removed) was from `2.0.305` to `2.0.310`. I was already using the branch `2.x` without problems.

wojtekmaj commented:
I think the misunderstanding is that usually on GitHub if something is pre-release it's marked as so, e.g. 2.0.305-beta or something like that. Mozilla just throws pdfjs-dist on npm "as is" with every merge, without e.g. releasing using a tag "next" or similar. So, if you take pdfjs-dist@latest you are actually getting the beta without knowing about it.

Snuffleupagus commented:
> I was already using the branch 2.x without problems.

That can probably be attributed to pure luck, given how you're currently using/integrating PDF.js in your application, rather than anything else.
As mentioned, PDF.js version `2.0` will be a breaking change w.r.t. to the API, and shouldn't be considered final until released officially at https://github.com/mozilla/pdf.js/releases.

Rob--W commented:
@wojtekmaj Your analysis is correct, we should improve the documentation. I filed #9440 for that.
